File: Examples/Python/PingPong.py
import sys
import System
import clr
clr.AddReference("MPI")
import MPI

args = System.Environment.GetCommandLineArgs()
argsref = clr.Reference[System.Array[str]](args)
env = MPI.Environment(argsref)

destHostname = ""
comm = MPI.Communicator.world
if comm.Rank == 0:
    print "Rank 0 is alive and running on " + MPI.Environment.ProcessorName
    for dest  in range(1, comm.Size):
        print "Pinging process with rank " + str(dest) + "..."
        comm.Send[str]("Ping!", dest, 0)
        destHostname = comm.Receive[str].Overloads[(int, int)](dest, 1)
        print " Pong!"
        print "  Rank " + str(dest) + " is alive and running on " + str(destHostname)
else:
    destHostname = comm.Receive[str].Overloads[(int, int)](0, 0)
    comm.Send[str](str(MPI.Environment.ProcessorName), 0, 1)

env.Dispose()

File: Examples/PingPong/PingPong.cs
/* Copyright (C) 2007  The Trustees of Indiana University
 *
 * Use, modification and distribution is subject to the Boost Software
 * License, Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
 * http://www.boost.org/LICENSE_1_0.txt)
 *  
 * Authors: Douglas Gregor
 *          Andrew Lumsdaine
 * 
 * This test exercises Communicator.ReduceScatter.
 */
using System;
using MPI;

class PingPong
{
    static void Main(string[] args)
    {
        MPI.Environment.Run(ref args, comm =>
        {
            if (comm.Rank == 0)
            {
                Console.WriteLine("Rank 0 is alive and running on " + MPI.Environment.ProcessorName);
                for (int dest = 1; dest < comm.Size; ++dest)
                {
                    Console.Write("Pinging process with rank " + dest + "...");
                    comm.Send("Ping!", dest, 0);
                    string destHostname = comm.Receive<string>(dest, 1);
                    Console.WriteLine(" Pong!");
                    Console.WriteLine("  Rank " + dest + " is alive and running on " + destHostname);
                }
            }
            else
            {
                comm.Receive<string>(0, 0);
                comm.Send(MPI.Environment.ProcessorName, 0, 1);
            }
        });
    }
}
